README.rst
Python wrapper for Telldus Live
===============================

.. image:: https://badge.fury.io/py/tellive-py.png
    :target: https://pypi.python.org/pypi/tellive-py/

.. image:: https://secure.travis-ci.org/erijo/tellive-py.png?branch=master
    :target: http://travis-ci.org/erijo/tellive-py

tellive-py is a Python wrapper for `Telldus Live <http://live.telldus.com/>`_,
"a user friendly service for automating your TellStick connected gear using the
Internet".

* Official home page: https://github.com/erijo/tellive-py
* Python package index: https://pypi.python.org/pypi/tellive-py

Please report any problem as a `GitHub issue report
<https://github.com/erijo/tellive-py/issues/new>`_.

Features
--------

* Includes the script `tellive_core_connector
  <https://github.com/erijo/tellive-py/blob/master/bin/tellive_core_connector>`_
  for connecting a e.g. a Tellstick Duo to Telldus Live without needing Telldus
  Center. Supports both devices and sensors.
* Open source (`GPLv3+
  <https://github.com/erijo/tellive-py/blob/master/LICENSE.txt>`_).

Requirements
------------

* Python 3.2+
* `tellcore-py <https://github.com/erijo/tellcore-py>`_
* On Mac OS X, `appnope <https://pypi.python.org/pypi/appnope>`_ is
  recommended.

Installation
------------

.. code-block:: bash

    $ pip install tellive-py

Example
-------

To run the included program for connecting a TellStick to Telldus Live:

.. code-block:: bash

    $ tellive_core_connector ~/.config/tellive.conf

The first time you run the program (with a particular config file), it will
exit and ask you to visit a given URL to give the program access to your
account.

When you have done so, you can then edit the config file
(``~/.config/tellive.conf`` in this example) and add a name to the sensors that
you wish to send to Telldus Live. You can also disable devices that you don't
want to be controllable via Telldus Live (see ``tellive_core_connector --help``
for more info). Then start the program again as above.

The API can also be used by your own program. This how you would connect to
Telldus Live and register the client (with PUBLIC_KEY and PRIVATE_KEY from
`here <http://api.telldus.com/keys/index>`_):

.. code-block:: python

    client = TellstickLiveClient(PUBLIC_KEY, PRIVATE_KEY)
    (server, port) = client.connect_to_first_available_server()
    client.register(version="0.1")


CHANGES.rst
Changelog
=========

0.5.2 (2014-11-25)
------------------

* Correctly handle last sent value when it is 0.


0.5.1 (2014-11-20)
------------------

* Don't try to start browser automatically during first run. In many cases it
  doesn't work and may hide the URL.
* Handle last_sent_value() returning None (issue #4).


0.5.0 (2014-11-19)
------------------

* Disable appnap on Mac OS X if appnope module is available (issue #2).
* Report new/changed/removed devices to Telldus live.
* Release socket(s) before waiting to re-connect.
* Require tellcore-py >= v1.1.0.


0.4.2 (2014-02-25)
------------------

* Fixed problem that could occur after disconnect from server.
* Fixed tellive_core_connector problem on Mac OS X (issue #1).


0.4.1 (2014-02-06)
------------------

* Add all sensors and devices to the config on the first run.


0.4.0 (2014-02-06)
------------------

* Fixed tellive_core_connector to not wake up two times every second, but
  instead only wake up when there is work to do.


0.3.0 (2014-02-04)
------------------

* Removed reload message as it is not supposed to be sent to clients.
* Better values for os and os-version in register message.
* Support marking devices as disabled to not show up in Telldus Live.


0.2.0 (2014-02-02)
------------------

* tellive_core_connector now uses official keys from Telldus, so you no longer
  need to use private tokens.
* Log using the standard logging module.
* Reconnect if connection is lost for some reason.
* Fixed problem with Python 3.2.
* Added support for reload request from server.
* Only report sensors that are named in the config file.


0.1.1 (2014-01-28)
------------------

* Fix some packaging issues.


0.1.0 (2014-01-28)
------------------

* Initial release.


tests/test_livemessage.py
# Copyright (c) 2014 Erik Johansson <erik@ejohansson.se>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA

import unittest

from tellive.livemessage import LiveMessageToken


class Test(unittest.TestCase):
    def assert_invalid(self, token):
        self.assertIs(token.value, None)

    def assert_int(self, token, expected):
        self.assertIs(type(token.value), int)
        self.assertEqual(token.value, expected)

    def assert_string(self, token, expected):
        self.assertIs(type(token.value), str)
        self.assertEqual(token.value, expected)

    def assert_base64(self, token, expected):
        self.assertIs(type(token.value), bytes)
        self.assertEqual(token.value, expected)

    def assert_list(self, token, expected):
        self.assertIs(type(token.value), list)
        self.assertListEqual(token.value, expected)

    def assert_dict(self, token, expected):
        self.assertIs(type(token.value), dict)
        self.assertDictEqual(token.value, expected)

    def test_empty_token(self):
        self.assert_invalid(LiveMessageToken())

    def test_int(self):
        token = LiveMessageToken(10)
        self.assert_int(token, 10)
        self.assertEqual(token.serialize(), b'iAs')

    def test_string(self):
        a_ring = b'\xc3\xa5'.decode('utf-8')
        token = LiveMessageToken("foobar123456" + a_ring)
        self.assert_string(token, "foobar123456" + a_ring)
        self.assertEqual(token.serialize(), b'E:foobar123456\xc3\xa5')

    def test_base64(self):
        token = LiveMessageToken(b'foobar123456')
        self.assert_base64(token, b'foobar123456')
        self.assertEqual(token.serialize(), b'u10:Zm9vYmFyMTIzNDU2')

    def test_empty_list(self):
        token = LiveMessageToken([])
        self.assert_list(token, [])
        self.assertEqual(token.serialize(), b'ls')

    def test_list(self):
        token = LiveMessageToken([1, "foo"])
        self.assert_list(token, [1, "foo"])
        self.assertEqual(token.serialize(), b'li1s3:foos')

    def test_empty_dict(self):
        token = LiveMessageToken({})
        self.assert_dict(token, {})
        self.assertEqual(token.serialize(), b'hs')

    def test_dict(self):
        token = LiveMessageToken({'1': "foo", '2': 3})
        self.assert_dict(token, {'1': "foo", '2': 3})
        # The order of keys are undefined
        serialized = token.serialize()
        if serialized != b'h1:13:foo1:2i3ss':
            self.assertEqual(serialized, b'h1:2i3s1:13:foos')

    def test_dict_in_list_in_dict(self):
        token = LiveMessageToken({'a': [{'b': 1}, 2]})
        self.assertIs(type(token.value['a']), list)
        self.assertEqual(token.serialize(), b'h1:alh1:bi1ssi2sss')

    def test_deserialize_empty(self):
        self.assertIs(None, LiveMessageToken.deserialize(b'')[0])

    def test_deserialize_invalid(self):
        self.assertIs(None, LiveMessageToken.deserialize(b'i')[0])
        self.assertIs(None, LiveMessageToken.deserialize(b'l')[0])
        self.assertIs(None, LiveMessageToken.deserialize(b'h')[0])
        self.assertIs(None, LiveMessageToken.deserialize(b'hi1s')[0])
        self.assertIs(None, LiveMessageToken.deserialize(b'u')[0])
        self.assertIs(None, LiveMessageToken.deserialize(b'u7:YWxpdmU')[0])

    def test_deserialize_int(self):
        (token, rest) = LiveMessageToken.deserialize(b'i1A2s')
        self.assert_int(token, 0x1a2)
        self.assertFalse(rest)

    def test_deserialize_empty_list(self):
        (token, rest) = LiveMessageToken.deserialize(b'ls')
        self.assert_list(token, [])
        self.assertFalse(rest)

    def test_deserialize_list(self):
        (token, rest) = LiveMessageToken.deserialize(b'li1siAss')
        self.assert_list(token, [1, 10])
        self.assertFalse(rest)

    def test_deserialize_empty_dict(self):
        (token, rest) = LiveMessageToken.deserialize(b'hs')
        self.assert_dict(token, {})
        self.assertFalse(rest)

    def test_deserialize_dict(self):
        (token, rest) = LiveMessageToken.deserialize(b'h1:13:foo1:26:foobars')
        self.assert_dict(token, {'1': "foo", '2': "foobar"})
        self.assertFalse(rest)

    def test_deserialize_dict_with_list(self):
        (token, rest) = LiveMessageToken.deserialize(
            b'h1:1li1Es1:Alss1:2h1:3hsss')
        self.assert_dict(token, {'1': [0x1e, "A", []], '2': {'3': {}}})
        self.assertFalse(rest)

    def test_deserialize_empty_base64(self):
        (token, rest) = LiveMessageToken.deserialize(b'u0:')
        self.assert_base64(token, b'')
        self.assertFalse(rest)

    def test_deserialize_base64(self):
        (token, rest) = LiveMessageToken.deserialize(b'u8:YWxpdmU=')
        self.assert_base64(token, b'alive')
        self.assertFalse(rest)

    def test_deserialize_empty_string(self):
        (token, rest) = LiveMessageToken.deserialize(b'0:')
        self.assert_string(token, "")
        self.assertFalse(rest)

    def test_deserialize_string(self):
        (token, rest) = LiveMessageToken.deserialize(b'10:123456789abcdefP')
        self.assert_string(token, "123456789abcdefP")
        self.assertFalse(rest)


if __name__ == '__main__':
    unittest.main()


tellive/live.py
# Copyright (c) 2014 Erik Johansson <erik@ejohansson.se>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA

import tellcore.constants as const

class TelldusLive(object):
    def __init__(self, client):
        super().__init__()
        self._client = client

    def devices(self, supported_methods=None):
        if supported_methods is None:
            supported_methods = const.TELLSTICK_TURNON \
                | const.TELLSTICK_TURNOFF \
                | const.TELLSTICK_BELL \
                | const.TELLSTICK_TOGGLE \
                | const.TELLSTICK_DIM \
                | const.TELLSTICK_LEARN \
                | const.TELLSTICK_EXECUTE \
                | const.TELLSTICK_UP \
                | const.TELLSTICK_DOWN \
                | const.TELLSTICK_STOP
        params = {'supportedMethos': supported_methods,
                  'extras': 'coordinate,timezone,tzoffset'}
        values = self._client.request("devices/list", params)
        return [Device(self._client, p['id'], p) for p in values['device']]


class Device(object):
    def __init__(self, client, id, params={}):
        super().__init__()
        super().__setattr__('_client', client)
        super().__setattr__('id', id)
        self._update(params)

    def _update(self, params):
        for name in ['name', 'state', 'statevalue']:
            if name in params:
                super().__setattr__(name, params[name])

    def __getattr__(self, name):
        for attempt in range(2):
            try:
                return self.__dict__[name]
            except KeyError:
                if attempt == 0:
                    self.refresh()
                    continue
                raise AttributeError(name) from None

    def refresh(self, supported_methods=None):
        if supported_methods is None:
            supported_methods = const.TELLSTICK_TURNON \
                | const.TELLSTICK_TURNOFF \
                | const.TELLSTICK_BELL \
                | const.TELLSTICK_TOGGLE \
                | const.TELLSTICK_DIM \
                | const.TELLSTICK_LEARN \
                | const.TELLSTICK_EXECUTE \
                | const.TELLSTICK_UP \
                | const.TELLSTICK_DOWN \
                | const.TELLSTICK_STOP
        params = {'id': self.id, 'supportedMethos': supported_methods,
                  'extras': 'coordinate,timezone,tzoffset'}
        values = self._client.request("device/info", params)
        self._update(values)

    def turn_on(self):
        params = {'id': self.id}
        self._client.request("device/turnOn", params)

    def turn_off(self):
        params = {'id': self.id}
        self._client.request("device/turnOff", params)

    def last_sent_command(self):
        return self.state

    def last_sent_value(self):
        return self.statevalue


tellive/client.py
# Copyright (c) 2014 Erik Johansson <erik@ejohansson.se>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA

import http.client as http
import json
import logging
import oauthlib.oauth1
import urllib.parse

class TelldusLiveError(Exception):
    def __init__(self, error):
        super().__init__(error)


class LiveClient(object):
    def __init__(self, public_key, private_key, server='api.telldus.com',
                 port=http.HTTP_PORT, access_token=None, access_secret=None):
        super().__init__()
        self.public_key = public_key
        self.private_key = private_key
        self.server = server
        self.port = port
        self.host = "http://{}".format(server)
        if port != http.HTTP_PORT:
            self.host += ":{}".format(port)
        self.token = access_token
        self.secret = access_secret

    def _request(self, path, token, secret):
        client = oauthlib.oauth1.Client(
            self.public_key, client_secret=self.private_key,
            resource_owner_key=token, resource_owner_secret=secret)

        uri, headers, body = client.sign(self.host + path)
        conn = http.HTTPConnection(self.server, self.port)
        conn.request('GET', path, body=body, headers=headers)

        response = conn.getresponse()
        if response.status != http.OK:
            raise RuntimeError(
                "Could not get {} from {}:{}: {} {}".format(
                    path, self.server, self.port, response.status,
                    response.reason))
        return response

    def _token(self, path, token=None, secret=None):
        response = self._request(path, token, secret)
        qs = urllib.parse.parse_qs(response.read().decode('utf-8'))
        return qs['oauth_token'][0], qs['oauth_token_secret'][0]

    def request_token(self):
        """ Returns url, request_token, request_secret"""
        logging.debug("Getting request token from %s:%d",
                      self.server, self.port)
        token, secret = self._token("/oauth/requestToken")
        return "{}/oauth/authorize?oauth_token={}".format(self.host, token), \
            token, secret

    def access_token(self, request_token, request_secret):
        """Returns access_token, access_secret"""
        logging.debug("Getting access token from %s:%d",
                      self.server, self.port)
        self.access_token, self.access_secret = \
            self._token("/oauth/accessToken", request_token, request_secret)
        return self.access_token, self.access_secret

    def request(self, method, params):
        path = "/json/{}?{}".format(method, urllib.parse.urlencode(params))
        response = self._request(path, self.token, self.secret)
        values = json.loads(response.read().decode('utf-8'))
        if 'error' in values:
            raise TelldusLiveError(values['error'])
        return values


tellive/livemessage.py
# Copyright (c) 2014 Erik Johansson <erik@ejohansson.se>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA

import base64
import hashlib


class LiveMessageToken(object):
    def __init__(self, value=None):
        super().__init__()

        if type(value) in (int, str, bytes, list, dict):
            self.value = value
        else:
            self.value = None

    def serialize(self):
        """Serialize the token and return it as bytes."""
        if type(self.value) == int:
            return "i{:X}s".format(self.value).encode('ascii')

        if type(self.value) == str:
            value = self.value.encode('utf-8')
            return "{:X}:".format(len(value)).encode('ascii') + value

        if type(self.value) == bytes:
            value = base64.standard_b64encode(self.value)
            return "u{:X}:".format(len(value)).encode('ascii') + value

        if type(self.value) == list:
            items = [LiveMessageToken(m).serialize() for m in self.value]
            return b'l' + b''.join(items) + b's'

        if type(self.value) == dict:
            items = []
            for key, value in self.value.items():
                items.append(LiveMessageToken(str(key)).serialize())
                items.append(LiveMessageToken(value).serialize())
            return b'h' + b''.join(items) + b's'

        raise RuntimeError("Unknown type %s" % type(self.value))

    @staticmethod
    def deserialize(data):
        def _find(data, needle):
            needle = ord(needle)
            index = 0
            while index < len(data):
                if data[index] == needle:
                    return index
                index += 1
            raise ValueError

        def deserialize_int(data):
            end = _find(data, 's')
            return (LiveMessageToken(int(data[:end], 16)), data[end + 1:])

        def deserialize_list(data):
            result = []
            while data[0] != ord('s'):
                token, data = LiveMessageToken.deserialize(data)
                result.append(token.value)
            return (LiveMessageToken(result), data[1:])

        def deserialize_dict(data):
            result = {}
            while data[0] != ord('s'):
                key, data = LiveMessageToken.deserialize(data)
                value, data = LiveMessageToken.deserialize(data)
                result[str(key.value)] = value.value
            return (LiveMessageToken(result), data[1:])

        def deserialize_string(data, is_base64=False):
            end = _find(data, ':')
            value_end = end + 1 + int(data[:end], 16)
            value = data[end + 1:value_end]
            if is_base64:
                value = base64.standard_b64decode(value)
            else:
                value = value.decode('utf-8')
            return (LiveMessageToken(value), data[value_end:])

        try:
            if data[0] == ord('i'):
                return deserialize_int(data[1:])
            elif data[0] == ord('l'):
                return deserialize_list(data[1:])
            elif data[0] == ord('h'):
                return deserialize_dict(data[1:])
            elif data[0] == ord('u'):
                return deserialize_string(data[1:], is_base64=True)
            else:
                return deserialize_string(data)
        except:
            return (None, None)


class LiveMessage(object):
    def __init__(self, subject=None):
        super().__init__()
        self.tokens = []
        if subject:
            self.append(subject)

    def append(self, value):
        self.tokens.append(value)

    def subject(self):
        return self.tokens[0].lower()

    def parameter(self, index):
        return self.tokens[index + 1]

    def create_signed_message(self, private_key, hash_method):
        data = self.serialize()
        signature = LiveMessage.signature(data, private_key, hash_method)

        envelope = LiveMessage(signature)
        envelope.append(data.decode('utf-8'))
        return envelope

    def verify_signature(self, private_key, hash_method):
        data = self.parameter(0).encode('utf-8')
        signature = LiveMessage.signature(data, private_key, hash_method)
        return self.subject() == signature

    def serialize(self):
        tokens = [LiveMessageToken(t).serialize() for t in self.tokens]
        return b''.join(tokens)

    @staticmethod
    def deserialize(data):
        message = LiveMessage()
        while data:
            token, data = LiveMessageToken.deserialize(data)
            if not token:
                break
            message.append(token.value)
        return message

    @staticmethod
    def signature(data, private_key, hash_method):
        if hash_method == "sha512":
            signature = hashlib.sha512()
        elif hash_method == "sha256":
            signature = hashlib.sha256()
        else:
            signature = hashlib.sha1()

        signature.update(data)
        signature.update(private_key.encode('ascii'))
        return signature.hexdigest().lower()


tellive/tellstick.py
# Copyright (c) 2014 Erik Johansson <erik@ejohansson.se>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA

from .livemessage import LiveMessage

import http.client as http
import logging
import platform
import socket
import ssl
import time
import xml.parsers.expat as expat


class TellstickLiveClient(object):
    # dict(supportedMethods)
    SUBJECT_REGISTERD = "registered"
    # dict(uuid, url)
    SUBJECT_NOT_REGISTERED = "notregistered"
    # dict(id, action, [value, ACK])
    SUBJECT_COMMAND = "command"
    # <no parameters>
    SUBJECT_PONG = "pong"
    # <no parameters>
    SUBJECT_DISCONNECT = "disconnect"

    def __init__(self, public_key, private_key):
        super().__init__()
        self.socket = None
        self.public_key = public_key
        self.private_key = private_key
        self.hash_method = "sha1"
        self.time_sent = 0
        self.time_received = 0

    def ssl_context(self):
        context = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
        context.verify_mode = ssl.CERT_REQUIRED
        context.set_default_verify_paths()
        return context

    def servers(self, server='api.telldus.com', port=http.HTTPS_PORT):
        """Fetch list of servers that can be connected to.

        :return: list of (address, port) tuples
        """
        logging.debug("Fetching server list from %s:%d", server, port)

        conn = http.HTTPSConnection(server, port, context=self.ssl_context())
        conn.request('GET', "/server/assign?protocolVersion=2")

        response = conn.getresponse()
        if response.status != http.OK:
            raise RuntimeError("Could not connect to {}:{}: {} {}".format(
                server, port, response.status, response.reason))

        servers = []

        def extract_servers(name, attributes):
            if name == "server":
                servers.append((attributes['address'],
                                int(attributes['port'])))

        parser = expat.ParserCreate()
        parser.StartElementHandler = extract_servers
        parser.ParseFile(response)

        logging.debug("Found %d available servers", len(servers))
        return servers

    def connect(self, address, timeout=5):
        sock = self.ssl_context().wrap_socket(
            socket.socket(socket.AF_INET, socket.SOCK_STREAM))
        sock.settimeout(timeout)
        sock.connect(address)
        ssl.match_hostname(sock.getpeercert(), address[0])
        self.socket = sock
        self.time_sent = time.time()
        self.time_received = self.time_sent

    def connect_to_first_available_server(self, **kwargs):
        for server in self.servers():
            try:
                logging.debug("Connecting to %s:%d", server[0], server[1])
                self.connect(server, **kwargs)
                return server
            except:
                pass
        raise RuntimeError("Could not connect to any available server")

    def disconnect(self):
        if self.socket:
            self.socket.close()
            self.socket = None

    def send_message(self, message):
        envelope = message.create_signed_message(
            self.private_key, self.hash_method)

        data = envelope.serialize()
        logging.debug("Sending: %s", data)
        self.socket.write(data)
        self.time_sent = time.time()

    def receive_message(self):
        data = self.socket.read(1024)
        logging.debug("Received: %s", data)
        self.time_received = time.time()

        envelope = LiveMessage.deserialize(data)
        if not envelope.verify_signature(self.private_key, self.hash_method):
            raise ValueError("Signature verification failed")

        return LiveMessage.deserialize(envelope.parameter(0).encode('utf-8'))

    def register(self, version, uuid=""):
        message = LiveMessage("Register")
        message.append({'key': self.public_key, 'uuid': uuid,
                        'hash': self.hash_method})
        os = platform.system().lower()
        if os == "linux":
            os_version = platform.linux_distribution()[0]
        elif os == "darwin":
            os = "macosx"
            os_version = platform.mac_ver()[0]
        elif os == "windows":
            os_version = platform.win32_ver()[0]
        else:
            os_version = ""
        message.append({'protocol': 2, 'version': str(version),
                        'os': os, 'os-version': os_version.lower()})
        self.send_message(message)

    def ping(self):
        message = LiveMessage("Ping")
        self.send_message(message)

    def acknowledge(self, cookie):
        message = LiveMessage("ACK")
        message.append(cookie)
        self.send_message(message)

    def report_devices(self, devices, supported_methods):
        dev_list = []
        for device in devices:
            dev = {'id': device.id, 'name': device.name}
            dev['methods'] = device.methods(supported_methods)
            dev['state'] = device.last_sent_command(supported_methods)
            last_sent = device.last_sent_value()
            dev['stateValue'] = str('' if last_sent is None else last_sent)
            dev_list.append(dev)
        message = LiveMessage("DevicesReport")
        message.append(dev_list)
        self.send_message(message)

    def report_device_event(self, device_id, method, data):
        message = LiveMessage("DeviceEvent")
        message.append(device_id)
        message.append(method)
        message.append(data)
        self.send_message(message)

    def _sensor(self, sensor):
        value_list = []
        for datatype in sensor.DATATYPES.values():
            if not sensor.has_value(datatype):
                continue
            value = sensor.value(datatype)
            value_list.append({'type': datatype, 'value': value.value,
                               'lastUp': value.timestamp})
            s = {'protocol': sensor.protocol, 'model': sensor.model,
                 'sensor_id': sensor.id}
        return s, value_list

    def report_sensors(self, sensors, name_function=None):
        if not name_function:
            name_function = lambda x: ''
        sensor_list = []
        for sensor in sensors:
            s, value_list = self._sensor(sensor)
            s['name'] = name_function(sensor)
            sensor_list.append([s, value_list])

        message = LiveMessage("SensorsReport")
        message.append(sensor_list)
        self.send_message(message)

    def report_sensor_values(self, sensor):
        s, value_list = self._sensor(sensor)
        message = LiveMessage("SensorEvent")
        message.append(s)
        message.append(value_list)
        self.send_message(message)
